mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -227,12 +243,139 @@ class LinearRegressionModel private[ml] (
   extends RegressionModel[Vector, LinearRegressionModel]
   with LinearRegressionParams {
 
+  @transient private var trainingSummary: Option[LinearRegressionTrainingSummary] = None
+
+  /**
+   * Gets results summary (e.g. residuals, mse, r-squared ) of model on training set. This method
+   * should only be called on the driver as `summary` is transient.
+   *
+   * An exception is thrown if `trainingSummary == None`.
+   */
+  def summary: LinearRegressionTrainingSummary = trainingSummary match {
+    case Some(summ) => summ
+    case None =>
+      throw new SparkException(
+        "No training summary available for this LinearRegressionModel",
+        new NullPointerException())
+  }
+
+  private[regression] def setSummary(summary: LinearRegressionTrainingSummary): this.type = {
+    this.trainingSummary = Some(summary)
+    this
+  }
+
+  def hasSummary: Boolean = trainingSummary.isDefined
+
+
+  /**
+   * Evaluates the model on a testset.
+   * @param dataset Test dataset to evaluate model on.
+   */
+  def evaluate(dataset: DataFrame): LinearRegressionSummary = {
+    val t = udf { features: Vector => predict(features) }
+    val predictionAndObservations = dataset
+      .select(col($(labelCol)), t(col($(featuresCol))).as($(predictionCol)))
+
+    new LinearRegressionSummary(predictionAndObservations, $(predictionCol), $(labelCol))
+  }
+
   override protected def predict(features: Vector): Double = {
     dot(features, weights) + intercept
   }
 
   override def copy(extra: ParamMap): LinearRegressionModel = {
-    copyValues(new LinearRegressionModel(uid, weights, intercept), extra)
+    val newModel = copyValues(new LinearRegressionModel(uid, weights, intercept))
+    if (trainingSummary.isDefined) newModel.setSummary(trainingSummary.get)
+    newModel
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Linear regression training results.
+ * @param predictions predictions outputted by the model's `transform` method.
+ * @param objectiveTrace objective function (scaled loss + regularization) at each iteration.
+ */
+@Experimental
+class LinearRegressionTrainingSummary private[ml] (
+    predictions: DataFrame,
+    predictionCol: String,
+    labelCol: String,
+    val objectiveTrace: DataFrame)
```

Comment:
This may be a problem.  If the goal is to save the model (+ summary) via Java serialization, I'm not sure how DataFrames will behave.  Even though they implement serializable, they should definitely not be used on workers.  We may need to rethink the previous design decision about representing non-scalar stats using DataFrames.